### PR TITLE
elk6: remove

### LINKED ISSUE #192342: Remove Elasticsearch, Logstash, Beats 6.x
The ELK stack 6.x went EOL in 2022/02.

Cf. https://endoflife.date/elasticsearch

cc @ajs124 who already removed kibana6.

### PR DESCRIPTION
**Pull request by ajs124**
###### Description of changes
Closes #192342

###### Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -6177,36 +6171,20 @@ with pkgs;
...
-    jre_headless = jre8_headless; # TODO: remove override https://github.com/NixOS/nixpkgs/pull/89731
-  };
   elasticsearch7 = callPackage ../servers/search/elasticsearch/7.x.nix {
     util-linux = util-linuxMinimal;
     jre_headless = jdk11_headless; # TODO: remove override https://github.com/NixOS/nixpkgs/pull/89731
   };
-  elasticsearch = elasticsearch6;
-  elasticsearch-oss = elasticsearch6-oss;
+  elasticsearch = elasticsearch7;
+  elasticsearch-oss = elasticsearch7-oss;
```
Comment (mweinelt) [inline pkgs/top-level/all-packages.nix:6181]: This seems to point nowhere.

File: pkgs/top-level/aliases.nix [hunk 2]
```diff
@@ -305,6 +305,7 @@ mapAliases ({
   ec2_api_tools = ec2-api-tools; # Added 2021-10-08
   ec2-utils = amazon-ec2-utils; # Added 2022-02-01
 
+  elasticsearch-oss = throw "elasticsearch-oss has been removed because nixpkgs currently only has the unfree version of elasticsearch7"; # Added 2022-10-04
```
Comment (mweinelt) [inline pkgs/top-level/aliases.nix:308]: Looks like there is no free variant anymore.

https://www.elastic.co/pricing/faq/licensing

**Review by mweinelt — APPROVED** [review]

**Review by mweinelt — CHANGES_REQUESTED** [review]
Oops, there are a few more `elasticsearch-oss` references out there.

```
pkgs/top-level/all-packages.nix
6181-
6182-  elasticsearchPlugins = recurseIntoAttrs (
6183-    callPackage ../servers/search/elasticsearch/plugins.nix {
6184:      elasticsearch = elasticsearch-oss;

nixos/tests/graylog.nix
8-
9-    services.mongodb.enable = true;
10-    services.elasticsearch.enable = true;
11:    services.elasticsearch.package = pkgs.elasticsearch-oss;

nixos/tests/parsedmarc/default.nix
84-            };
85-          };
86-
87:          services.elasticsearch.package = pkgs.elasticsearch-oss;
--
159-                };
160-              };
161-
162:              services.elasticsearch.package = pkgs.elasticsearch-oss;

**Comment by mweinelt** [comment]
@ofborg test parsedmarc graylog

**Comment by mweinelt** [comment]
cc @talyz @Ma27 

Given that elasticsearch7 is unfree this breaks in the nixos test runner for graylog and parsedmarc. Any ideas?

**Comment by dpausp** [comment]
The Graylog package is still at 3.3 and their docs say that it only works with Elasticsearch 6. Upgrading to Graylog 4.0 wouldn't help because it supports ES 7 up to 7.10.x which is already gone. 

We still use Graylog at the Flying Circus and we would prefer keeping Graylog 3.3 and ES 6 in for 22.11. We will have to maintain it anyway and can do that also in nixpkgs, if needed. We don't want to upgrade to Graylog 4 and will probably get rid of Graylog altogether next year, their change to SSPL being one of the reasons.

Another solution for the future would be to upgrade to Graylog 4.3 and use OpenSearch. I already have an experimental OpenSearch module running and want to work on it next week.

**Comment by yorickvP** [comment]
@dpausp I've been looking at upgrading graylog in nixpkgs to 4.3 because of CVE-2021-37759 [9.8] and CVE-2021-37760 [9.8].

Maybe it's worth keeping it around but marked insecure?

**Comment by talyz** [comment]
OpenSearch support is probably the way forward for parsedmarc as well. Currently, they've just pinned the old version of the `elasticsearch` python package to maintain compatibility - one alternative would be for us to do the same once we have an OpenSearch module in place.

**Comment by mweinelt** [comment]
Anyone willing to package up OpenSearch then? How about Graylog 4.3? Let's not drop the ball here.

**Comment by dpausp** [comment]
> Anyone willing to package up OpenSearch then? How about Graylog 4.3? Let's not drop the ball here.

I have OpenSearch 2.3.0 running here, draft PR will follow later today.

**Comment by dpausp** [comment]
First working draft of OpenSearch at #197800, who wants to try Graylog 4.3? with it? :)

**Comment by mweinelt** [comment]
Postponed for 23.05. Please work on the outstanding issues, I'm not inclined to keep networked software around for well past their EOL.

**Comment by mweinelt** [comment]
OpenSearch incoming at https://github.com/NixOS/nixpkgs/pull/175000.

**Comment by mweinelt** [comment]
@dpausp Both ES6 and ES7 will not be part of 23.05, so please look into how you can keep graylog working, if you still rely on that.

We will not delay this backport for another release, when the issue was well known.

**Comment by mweinelt** [comment]
Let's also remove elasticsearch7, it is going to EOL two months into the 23.05 release. https://endoflife.date/elasticsearch

For graylog someone would need to update to the 5.x series, which supports opensearch 1.x, which would also need to be packaged. But at least that is supported until 2023/12. https://endoflife.date/opensearch

for graylog @fadenb @dpausp 
for elasticsearch7 @basvandijk @apeschar

**Comment by dpausp** [comment]
> @dpausp Both ES6 and ES7 will not be part of 23.05, so please look into how you can keep graylog working, if you still rely on that.
> 
> We will not delay this backport for another release, when the issue was well known.

ACK, we decided to drop graylog in favor of Loki for the Flying Circus platform and I don't have the time to keep graylog working myself.

**Comment by Artturin** [comment]
```patch
diff --git a/pkgs/servers/search/elasticsearch/plugins.nix b/pkgs/servers/search/elasticsearch/plugins.nix
index cb595de6a13..616a85f7fdd 100644
--- a/pkgs/servers/search/elasticsearch/plugins.nix
+++ b/pkgs/servers/search/elasticsearch/plugins.nix
@@ -39,7 +39,6 @@ in
       url = "https://artifacts.elastic.co/downloads/elasticsearch-plugins/${pluginName}/${pluginName}-${version}.zip";
       sha256 =
         if version == "7.17.4" then "a4e881d86694ae70ab6b18f72ea700415971200145d33d438e57c0374d9fc16f"
-        else if version == "6.8.21" then "06b1pavyggzfp4wwdql0q9nm3r7i9px9cagp4yh4nhxhnk4w5fiq"
         else throw "unsupported version ${version} for plugin ${pluginName}";
     };
     meta = with lib; {
@@ -56,7 +55,6 @@ in
       url = "https://github.com/vhyza/elasticsearch-${pluginName}/releases/download/v${version}/elasticsearch-${pluginName}-${version}-plugin.zip";
       sha256 =
         if version == "7.17.3" then "1835f374230cb17193859cee22ac90e3d7a67fb41a55fd4578e840d708287a08"
-        else if version == "6.8.21" then "0m80cn7vkcvk95v4pdmi6vk5ww7p01k0hj2iqb9g870vs6x2qjzv"
         else throw "unsupported version ${version} for plugin ${pluginName}";
     };
     meta = with lib; {
@@ -73,7 +71,6 @@ in
       url = "https://artifacts.elastic.co/downloads/elasticsearch-plugins/${pluginName}/${pluginName}-${version}.zip";
       sha256 =
         if version == "7.17.4" then "1c8175b2dac54277c1f41981fb4a784829e74e6e74268381fe0c27bc6652704b"
-        else if version == "6.8.21" then "07w8s4a5gvr9lzjzf629y8rx3kvs6zd1vl07ksw1paghp42yb354"
         else throw "unsupported version ${version} for plugin ${pluginName}";
     };
     meta = with lib; {
@@ -90,7 +87,6 @@ in
       url = "https://artifacts.elastic.co/downloads/elasticsearch-plugins/${pluginName}/${pluginName}-${version}.zip";
       sha256 =
         if version == "7.17.4" then "702e446997bde5cb38af120a1cb4271d976fdd23444be49e53b6be3801d845a9"
-        else if version == "6.8.21" then "1kdpbrasxwr3dn21zjrklp1s389rwa51fairygdwl8px9liwwfa5"
         else throw "unsupported version ${version} for plugin ${pluginName}";
     };
     meta = with lib; {
@@ -107,7 +103,6 @@ in
       url = "https://artifacts.elastic.co/downloads/elasticsearch-plugins/${pluginName}/${pluginName}-${version}.zip";
       sha256 =
         if version == "7.17.4" then "7d1574a585a9db0988ee248159d51f62cce5578a8c082096ef3e26efdb24aee7"
-        else if version == "6.8.21" then "0v31yyhjcdlqnjw1f9kihh7z3c6d31whc57hqqd1dn579n4s9rlz"
         else throw "unsupported version ${version} for plugin ${pluginName}";
     };
     meta = with lib; {
@@ -124,7 +119,6 @@ in
       url = "https://artifacts.elastic.co/downloads/elasticsearch-plugins/${pluginName}/${pluginName}-${esVersion}.zip";
       sha256 =
         if version == "7.17.4" then "cad923a662db705d40ca29698aa118e9e4cc50ae564c426a76d5acb777a4f57c"
-        else if version == "6.8.21" then "0sfh1az30q4f34zxig2fz8wn9gk53fmmxyg5pbi1svn9761p5awq"
         else throw "unsupported version ${version} for plugin ${pluginName}";
     };
     meta = with lib; {
@@ -141,7 +135,6 @@ in
       url = "https://artifacts.elastic.co/downloads/elasticsearch-plugins/${pluginName}/${pluginName}-${esVersion}.zip";
       sha256 =
         if version == "7.17.4" then "a50be4cea5c68ad7615f87d672ba160d027fdfde2be0578bb2dabd6384cc8108"
-        else if version == "6.8.21" then "00lwj00rfdk6850gk1n86chiz2w6afpqn7jn588jdbwv41qh5mrv"
         else throw "unsupported version ${version} for plugin ${pluginName}";
     };
     meta = with lib; {
@@ -158,7 +151,6 @@ in
     version =
       # https://docs.search-guard.com/latest/search-guard-versions
       if esVersion == "7.17.3" then "${esVersion}-53.1.0"
-      else if esVersion == "6.8.21" then "${esVersion}-25.6"
       else throw "unsupported version ${esVersion} for plugin ${pluginName}";
     src =
       if esVersion == "7.17.3" then
@@ -166,11 +158,6 @@ in
           url = "https://maven.search-guard.com/search-guard-suite-release/com/floragunn/search-guard-suite-plugin/${version}/search-guard-suite-plugin-${version}.zip";
           sha256 = "b49b24f7b74043cb5bab93f18316ea71656a7668e61bf063ccaa7b0ee2302a31";
         }
-      else if esVersion == "6.8.21" then
-        fetchurl {
-          url = "https://maven.search-guard.com/search-guard-release/com/floragunn/search-guard-6/${version}/search-guard-6-${version}.zip";
-          sha256 = "19nj513wigwd0mzq747zax4fzvv5vi24f7j0636rydd9iv9cyhg2";
-        }
       else throw "unsupported version ${version} for plugin ${pluginName}";
     meta = with lib; {
       homepage = "https://search-guard.com";
```

**Pull request merged**